src/OpenTelemetry.Adapter.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Adapter.Dependencies.Implementation
+{
+    internal class GrpcClientDiagnosticListener : ListenerHandler
+    {
+        private readonly PropertyFetcher startRequestFetcher = new PropertyFetcher("Request");
+
+        public GrpcClientDiagnosticListener(Tracer tracer)
+            : base("Grpc.Net.Client", tracer)
+        {
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStartActivity";
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            {
+                AdapterEventSource.Log.NullPayload(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            if (request.Headers.Contains("traceparent"))
+            {
+                // this request is already instrumented, we should back off
+                return;
+            }
+
+            // The path is also available from the grpc.method tag on the activity. Maybe that should be used instead
+            var path = request.RequestUri.AbsolutePath.Trim('/');
+            this.Tracer.StartActiveSpanFromActivity(path, activity, SpanKind.Client, out var span);
+
+            if (span.IsRecording)
+            {
+                var rx = new Regex(@"(?<package>\w+).(?<service>\w+)/(?<method>\w+)", RegexOptions.Compiled);
+                var rpcService = rx.Match(path).Groups["service"].Value;
+
+                span.SetAttribute("rpc.service", rpcService);
+
+                var uriHostNameType = Uri.CheckHostName(request.RequestUri.Host);
+                if (uriHostNameType == UriHostNameType.IPv4 || uriHostNameType == UriHostNameType.IPv6)
+                {
+                    span.SetAttribute("net.peer.ip", request.RequestUri.Host);
+                }
+                else
+                {
+                    span.SetAttribute("net.peer.name", request.RequestUri.Host);
+                }
+
+                span.SetAttribute("net.peer.port", request.RequestUri.Port);
+            }
+
+            // Do we want to propagate trace context here? GrpcClient is built on top of HttpClient.
```

Comment:
I'd like to understand this better. Do we end up with 2 spans, one from httpclient, and another from grpcclient, both essentially representing the same thing?

Reply:
Correct, two spans are generated. First the one from the grpcclient and then one from httpclient as a child.

I think it makes sense to only produce the span from the grpcclient. HttpClient instrumentation [checks for the traceparent header](https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/src/OpenTelemetry.Instrumentation.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs#L70-L74) which seems to suggest this kind of scenario was anticipated? Though `traceparent` assumes W3C format, so it seems like there needs to be a more general check here if we want to prevent both spans from being generated.

Reply:
The actual order of events would be like Grpc.Start, HttpClient.Start, HttpClient.Stop, Grpc.Stop right? The httpclient span would be the child of grpc one.

This is essentially same as a user making a httpclient call inside a using(span=startactivespan()) block, which result in 2 span, 1 from customer created, and other from httpclient. In this case, we'd want to see both spans. customer span as parent, httpclient span as child.

Applying this principle to Grpc->HttpClient, it looks okay to see both spans. But if both are essentially same information then its redundant. Maybe let user decide if they want to suppress the child.?

Instead of looking at traceparent presence, we can check the activity.Parent inside httpclient, which would be that of Grpc.

(Just my initial thoughts, we need to discuss this more)

Reply:
> The actual order of events would be like Grpc.Start, HttpClient.Start, HttpClient.Stop, Grpc.Stop right? The httpclient span would be the child of grpc one.

Correct.

My initial impression was that the information was redundant, though I can understand why it may be desirable to show both since it is an accurate representation of what's really happening.

I've seen similar cases from some database drivers that use HttpClient or HttpWebRequest behind the scenes. I've found it intuitive to show only a span for the database operation but not the underlying http request.

I'll share some screenshots of some traces in the gitter channel tomorrow, so we can discuss further.

Reply:
Okay.
Adding @shirhatti to this discussion :)

Reply:
Adding @jamesnk  as well.

Reply:
I don't know. Having both would be messy, so I like having just the gRPC span, but I don't know what scenarios would be impacted by not having the HttpClient span show up.

Reply:
I'm curious what the HttpClient activity lifetime is when making a gRPC streaming call. Does the HttpClient activity start with the request being sent and end with the request headers being received, or does it last until the response is complete?

Reply:
The traces I've generated seem to suggest that the HttpClient activity ends with the request headers being received. I can dig into the code to be sure of this. Though here's a trace I've captured:

<img width="1053" alt="Screen Shot 2020-05-22 at 6 18 39 PM" src="https://user-images.githubusercontent.com/3676547/82718680-d1416380-9c58-11ea-8d96-9bab656ff0e7.png">

Reply:
I agree and I think having both HTTP and gRPC spans is a little untidy, however I do think it's important to indicate the gPRC call was made over HTTP (it is theoretically possible to use other transports than HTTP).

It may also help identify an issue with the HTTP client when trying to help diagnose an issue by highlighting extra latency / errors.

Reply:
> it is theoretically possible to use other transports than HTTP

Ah yes, interesting. I had not thought about this use case. Agreed it would be useful to note that the call was over HTTP in this case. As far as I know, the library currently only supports HTTP/2 at the moment (perhaps I'm wrong about this?)

Reply:
@cijothomas and I spoke briefly regarding this issue. His thoughts:

> my take would be this - lets not make any assumptions about what user wants - make it configurable to user. let user decide if they want to see grpc only or httpclient only or both. we can discuss the default behavior later.

I think this stance makes sense for now. My thought for this PR is to move forward without any attempt to suppress the spans generated from the HttpClient instrumentation. Then perhaps follow up as a separate effort with the idea of making this configurable and deciding on defaults.

Reply:
I agree and think we should show everything as you are now. If we need to tighten up later, we can, maybe though configuration.